File: runtime-c/src/lexer.cpp
#include "lexer.h"

namespace evc {

namespace {

// Decode a UTF-8 byte string into Unicode codepoints. Evident operators (∈, ∧,
// ⇒, …) are single codepoints, so working over codepoints (like Rust's
// `chars()`) makes the operator matching below a direct mirror of lexer.rs.
std::vector<char32_t> decode_utf8(const std::string &src) {
    std::vector<char32_t> out;
    out.reserve(src.size());
    size_t i = 0, n = src.size();
    while (i < n) {
        unsigned char c = (unsigned char)src[i];
        char32_t cp;
        int len;
        if (c < 0x80) { cp = c; len = 1; }
        else if ((c >> 5) == 0x6) { cp = c & 0x1F; len = 2; }
        else if ((c >> 4) == 0xE) { cp = c & 0x0F; len = 3; }
        else if ((c >> 3) == 0x1E) { cp = c & 0x07; len = 4; }
        else { cp = 0xFFFD; len = 1; }
        for (int k = 1; k < len && i + k < n; k++)
            cp = (cp << 6) | ((unsigned char)src[i + k] & 0x3F);
        out.push_back(cp);
        i += len;
    }
    return out;
}

// Re-encode a single codepoint as UTF-8 bytes appended to `out`.
void encode_utf8(char32_t cp, std::string &out) {
    if (cp < 0x80) {
        out.push_back((char)cp);
    } else if (cp < 0x800) {
        out.push_back((char)(0xC0 | (cp >> 6)));
        out.push_back((char)(0x80 | (cp & 0x3F)));
    } else if (cp < 0x10000) {
        out.push_back((char)(0xE0 | (cp >> 12)));
        out.push_back((char)(0x80 | ((cp >> 6) & 0x3F)));
        out.push_back((char)(0x80 | (cp & 0x3F)));
    } else {
        out.push_back((char)(0xF0 | (cp >> 18)));
        out.push_back((char)(0x80 | ((cp >> 12) & 0x3F)));
        out.push_back((char)(0x80 | ((cp >> 6) & 0x3F)));
        out.push_back((char)(0x80 | (cp & 0x3F)));
    }
}

bool is_ident_start(char32_t c) {
    return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_';
}
bool is_ident_continue(char32_t c) {
    return is_ident_start(c) || (c >= '0' && c <= '9');
}
bool is_digit(char32_t c) { return c >= '0' && c <= '9'; }

Token keyword_or_ident(const std::string &s) {
    using K = Token::Kind;
    if (s == "schema")   return Token(K::Schema);
    if (s == "claim")    return Token(K::Claim);
    if (s == "type")     return Token(K::Type);
    if (s == "subclaim") return Token(K::Subclaim);
    if (s == "fsm")      return Token(K::Fsm);
    if (s == "external") return Token(K::External);
    if (s == "enum")     return Token(K::Enum);
    if (s == "match")    return Token(K::Match);
    if (s == "matches")  return Token(K::Matches);
    if (s == "import")   return Token(K::Import);
    if (s == "in")       return Token(K::In);
    if (s == "true")     return Token(K::True);
    if (s == "false")    return Token(K::False);
    if (s == "mapsto")   return Token(K::MapsTo);
    Token t(K::Ident);
    t.str = s;
    return t;
}

}  // namespace

std::vector<Token> tokenize(const std::string &src) {
    using K = Token::Kind;
    std::vector<char32_t> cp = decode_utf8(src);
    std::vector<Token> toks;

    size_t i = 0, n = cp.size();
    size_t line = 1, col = 1;
    bool at_line_start = true;
    size_t paren_depth = 0;

    auto peek = [&](size_t off = 0) -> char32_t {
        return (i + off < n) ? cp[i + off] : 0;
    };

    while (i < n) {
        char32_t c = cp[i];

        if (at_line_start) {
            size_t indent = 0;
            while (i < n) {
                if (cp[i] == ' ') { i++; col++; indent++; }
                else if (cp[i] == '\t') { i++; col++; indent += 4; }
                else break;
            }
            if (i >= n) break;  // EOF after indent
            char32_t ch = cp[i];
            if (ch == '\n') { i++; line++; col = 1; at_line_start = true; continue; }
            if (ch == '-' && peek(1) == '-') {
                while (i < n && cp[i] != '\n') { i++; col++; }
                continue;  // comment line: still at_line_start, newline handled next loop
            }
            Token t(K::Indent);
            t.indent = indent;
            toks.push_back(t);
            at_line_start = false;
            continue;
        }

        switch (c) {
            case ' ': case '\t': i++; col++; break;
            case '\n':
                i++; line++; col = 1;
                if (paren_depth == 0) { toks.push_back(Token(K::Newline)); at_line_start = true; }
                break;
            case '-':
                if (peek(1) == '-') {
                    while (i < n && cp[i] != '\n') { i++; col++; }
                } else { i++; col++; toks.push_back(Token(K::Minus)); }
                break;
            case '"': {
                i++; col++;
                std::string s;
                bool closed = false;
                while (i < n) {
                    char32_t d = cp[i];
                    if (d == '"') { i++; col++; closed = true; break; }
                    if (d == '\\') {
                        i++; col++;
                        char32_t e = peek();
                        switch (e) {
                            case '"':  s.push_back('"');  i++; col++; break;
                            case '\\': s.push_back('\\'); i++; col++; break;
                            case 'n':  s.push_back('\n'); i++; col++; break;
                            case 't':  s.push_back('\t'); i++; col++; break;
                            case 0: throw LexError("unterminated string escape", line, col);
                            default: throw LexError("unknown escape", line, col);
                        }
                        continue;
                    }
                    if (d == '\n') throw LexError("unterminated string literal", line, col);
                    encode_utf8(d, s);
                    i++; col++;
                }
                if (!closed) throw LexError("unterminated string at EOF", line, col);
                Token t(K::Str); t.str = s; toks.push_back(t);
                break;
            }
            default:
                if (is_digit(c)) {
                    std::string s;
                    while (i < n && is_digit(cp[i])) { s.push_back((char)cp[i]); i++; col++; }
                    // Real: digits '.' digits — only when a digit follows the dot.
                    if (peek() == '.' && is_digit(peek(1))) {
                        i++; col++; s.push_back('.');
                        while (i < n && is_digit(cp[i])) { s.push_back((char)cp[i]); i++; col++; }
                        Token t(K::Real); t.rval = std::stod(s); toks.push_back(t);
                    } else {
                        Token t(K::Int); t.ival = std::stoll(s); toks.push_back(t);
                    }
                } else if (is_ident_start(c)) {
                    std::string s;
                    while (i < n && is_ident_continue(cp[i])) { s.push_back((char)cp[i]); i++; col++; }
                    toks.push_back(keyword_or_ident(s));
                } else {
                    // single/multi-char operators
                    switch (c) {
                        case '+':
                            i++; col++;
                            if (peek() == '+') { i++; col++; toks.push_back(Token(K::PlusPlus)); }
                            else toks.push_back(Token(K::Plus));
                            break;
                        case '*': i++; col++; toks.push_back(Token(K::Star)); break;
                        case '/': i++; col++; toks.push_back(Token(K::Slash)); break;
                        case '(': i++; col++; toks.push_back(Token(K::LParen)); paren_depth++; break;
                        case ')': i++; col++; toks.push_back(Token(K::RParen)); if (paren_depth) paren_depth--; break;
                        case '{': i++; col++; toks.push_back(Token(K::LBrace)); paren_depth++; break;
                        case '}': i++; col++; toks.push_back(Token(K::RBrace)); if (paren_depth) paren_depth--; break;
                        case '[': i++; col++; toks.push_back(Token(K::LBracket)); paren_depth++; break;
                        case ']': i++; col++; toks.push_back(Token(K::RBracket)); if (paren_depth) paren_depth--; break;
                        case '#': i++; col++; toks.push_back(Token(K::Hash)); break;
                        case ',': i++; col++; toks.push_back(Token(K::Comma)); break;
                        case ':': i++; col++; toks.push_back(Token(K::Colon)); break;
                        case '.':
                            i++; col++;
                            if (peek() == '.') { i++; col++; toks.push_back(Token(K::DotDot)); }
                            else toks.push_back(Token(K::Dot));
                            break;
                        case '=':
                            i++; col++;
                            if (peek() == '>') { i++; col++; toks.push_back(Token(K::Implies)); }
                            else toks.push_back(Token(K::Eq));
                            break;
                        case '<':
                            i++; col++;
                            if (peek() == '=') { i++; col++; toks.push_back(Token(K::Le)); }
                            else toks.push_back(Token(K::Lt));
                            break;
                        case '>':
                            i++; col++;
                            if (peek() == '=') { i++; col++; toks.push_back(Token(K::Ge)); }
                            else toks.push_back(Token(K::Gt));
                            break;
                        case '!':
                            i++; col++;
                            if (peek() == '=') { i++; col++; toks.push_back(Token(K::Neq)); }
                            else throw LexError("unexpected '!'", line, col);
                            break;
                        case '|': i++; col++; toks.push_back(Token(K::Pipe)); break;
                        case '?': i++; col++; toks.push_back(Token(K::Question)); break;
                        // Unicode operators (codepoints).
                        case 0x2208: i++; col++; toks.push_back(Token(K::In)); break;          // ∈
                        case 0x2209: i++; col++; toks.push_back(Token(K::NotIn)); break;        // ∉
                        case 0x220B: i++; col++; toks.push_back(Token(K::ContainsRev)); break;  // ∋
                        case 0x2227: i++; col++; toks.push_back(Token(K::And)); break;          // ∧
                        case 0x2228: i++; col++; toks.push_back(Token(K::Or)); break;           // ∨
                        case 0x00AC: i++; col++; toks.push_back(Token(K::Not)); break;          // ¬
                        case 0x21D2: i++; col++; toks.push_back(Token(K::Implies)); break;      // ⇒
                        case 0x2264: i++; col++; toks.push_back(Token(K::Le)); break;           // ≤
                        case 0x2265: i++; col++; toks.push_back(Token(K::Ge)); break;           // ≥
                        case 0x2260: i++; col++; toks.push_back(Token(K::Neq)); break;          // ≠
                        case 0x2200: i++; col++; toks.push_back(Token(K::ForAll)); break;       // ∀
                        case 0x2203: i++; col++; toks.push_back(Token(K::Exists)); break;       // ∃
                        case 0x21A6: i++; col++; toks.push_back(Token(K::MapsTo)); break;       // ↦
                        case 0x27E8: i++; col++; toks.push_back(Token(K::LSeq)); paren_depth++; break;  // ⟨
                        case 0x27E9: i++; col++; toks.push_back(Token(K::RSeq)); if (paren_depth) paren_depth--; break;  // ⟩
                        default:
                            throw LexError("unexpected character (codepoint " +
                                           std::to_string((uint32_t)c) + ")", line, col);
                    }
                }
        }
    }

    toks.push_back(Token(K::Eof));
    return toks;
}

}  // namespace evc

File: runtime/src/lexer.rs
/// One segment inside an f-string literal.
#[derive(Debug, Clone, PartialEq)]
pub enum FStrPart {
    /// A run of literal characters between (or around) interpolations.
    Lit(String),
    /// The raw source text of a `{expr}` interpolation (without braces).
    Interp(String),
}

#[derive(Debug, Clone, PartialEq)]
pub enum Token {

    Ident(String),
    Int(i64),
    Real(f64),
    Str(String),
    /// f"..." format-string literal: alternating literal/interp parts.
    FStr(Vec<FStrPart>),
    True,
    False,

    Schema,
    Claim,
    Type,
    Subclaim,
    Operator,
    Fsm,
    Fti,
    External,
    Enum,
    Match,
    Matches,
    Import,
    In,
    NotIn,
    ContainsRev,

    Eq,
    Neq,
    Lt,
    Le,
    Gt,
    Ge,
    Plus,
    PlusPlus,
    Minus,
    Star,
    Slash,
    MidDot,
    Times,

    And,
    Or,
    Not,
    Implies,
    Delta,

    LParen,
    RParen,
    LBrace,
    RBrace,
    LBracket,
    RBracket,
    LSeq,
    RSeq,
    Hash,
    Comma,
    Pipe,
    Question,
    DotDot,
    Dot,
    Colon,
    ColonEq,
    ForAll,
    Exists,
    MapsTo,

    Newline,

    Indent(usize),

    Eof,
}

#[derive(Debug, Clone)]
pub struct LexError {
    pub message: String,
    pub line: usize,
    pub col: usize,
}

impl std::fmt::Display for LexError {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        write!(f, "lex error at line {}, col {}: {}", self.line, self.col, self.message)
    }
}

impl std::error::Error for LexError {}

/// Does the token stream end on an infix operator that REQUIRES a right operand? If so a
/// trailing newline is a line CONTINUATION, not a statement terminator — so `… ++` ⏎ `…` lexes
/// as one expression. Limited to ARITHMETIC and LOGICAL infix operators, which unambiguously
/// need a right operand and have NO block form. Deliberately EXCLUDED: `=` (and comparisons /
/// `↦`) — `=` ends the head of a multi-line `enum X =` ⏎ <variants> / `type X =` and a
/// chained-membership decl, where the trailing newline is load-bearing; `⇒`/`:` have block
/// forms too (implies-block, ternary). `¬` is prefix, and `,`/brackets are handled by paren
/// depth — none belong here.
fn ends_with_continuation_op(tokens: &[Token]) -> bool {
    matches!(
        tokens.last(),
        Some(
            Token::PlusPlus | Token::Plus | Token::Minus | Token::Star | Token::Slash
                | Token::MidDot | Token::Times | Token::And | Token::Or
        )
    )
}

/// Tokenize, dropping position info. Thin wrapper over [`tokenize_with_locs`].
pub fn tokenize(src: &str) -> Result<Vec<Token>, LexError> {
    tokenize_with_locs(src).map(|(toks, _locs)| toks)
}

/// Tokenize, returning a `(line, col)` position parallel to each token
/// (same index; the trailing `Eof` carries the final position). The parser
/// uses these to stamp parse errors with the offending token's location.
pub fn tokenize_with_locs(src: &str) -> Result<(Vec<Token>, Vec<(usize, usize)>), LexError> {
    let mut tokens = Vec::new();
    let mut locs: Vec<(usize, usize)> = Vec::new();
    let mut chars = src.chars().peekable();
    let mut line = 1usize;
    let mut col = 1usize;

    let mut at_line_start = true;
    let mut current_indent;

    let mut paren_depth: usize = 0;

    // After each scan step we backfill `locs` so it stays index-parallel with
    // `tokens`: the position recorded is where that token *started*.
    macro_rules! sync_locs {
        ($pos:expr) => {
            while locs.len() < tokens.len() {
                locs.push($pos);
            }
        };
    }

    while let Some(&c) = chars.peek() {
        let tok_start = (line, col);
        if at_line_start {

            current_indent = 0;
            while let Some(&ch) = chars.peek() {
                match ch {
                    ' '  => { chars.next(); col += 1; current_indent += 1; }
                    '\t' => { chars.next(); col += 1; current_indent += 4; }
                    _    => break,
                }
            }

            if let Some(&ch) = chars.peek() {
                if ch == '\n' {
                    chars.next(); line += 1; col = 1;
                    at_line_start = true;
                    continue;
                }
                if ch == '-' {

                    let mut clone = chars.clone();
                    clone.next();
                    if clone.peek() == Some(&'-') {

                        while let Some(&ch) = chars.peek() {
                            if ch == '\n' { break; }
                            chars.next(); col += 1;
                        }
                        continue;
                    }
                }
            } else {

                break;
            }
            tokens.push(Token::Indent(current_indent));
            sync_locs!(tok_start);
            at_line_start = false;
            continue;
        }

        match c {
            ' ' | '\t' => { chars.next(); col += 1; }
            '\n' => {
                chars.next();
                line += 1; col = 1;
                // A line ending on an infix operator that REQUIRES a right operand (`++`, `+`,
                // `∧`, a comparison, …) continues onto the next line — suppress the Newline so a
                // multi-line `"a" ++ "b" ++` ⏎ `"c"` lexes as one expression. `⇒` is deliberately
                // NOT in the set: `cond ⇒` ⏎ <indent> is the valid implies-block form.
                if paren_depth == 0 && !ends_with_continuation_op(&tokens) {
                    tokens.push(Token::Newline);
                    at_line_start = true;
                }

            }
            '-' => {

                let mut clone = chars.clone();
                clone.next();
                if clone.peek() == Some(&'-') {

                    while let Some(&ch) = chars.peek() {
                        if ch == '\n' { break; }
                        chars.next(); col += 1;
                    }
                } else {
                    chars.next(); col += 1;
                    tokens.push(Token::Minus);
                }
            }
            '"' => {

                chars.next(); col += 1;
                let mut s = String::new();
                loop {
                    match chars.peek().copied() {
                        Some('"') => { chars.next(); col += 1; break; }
                        Some('\\') => {
                            chars.next(); col += 1;
                            match chars.peek().copied() {
                                Some('"')  => { s.push('"');  chars.next(); col += 1; }
                                Some('\\') => { s.push('\\'); chars.next(); col += 1; }
                                Some('n')  => { s.push('\n'); chars.next(); col += 1; }
                                Some('t')  => { s.push('\t'); chars.next(); col += 1; }
                                Some(c)    => return Err(LexError {
                                    message: format!("unknown escape \\{}", c), line, col }),
                                None       => return Err(LexError {
                                    message: "unterminated string escape".into(), line, col }),
                            }
                        }
                        Some('\n') => return Err(LexError {
                            message: "unterminated string literal".into(), line, col }),
                        Some(c) => { s.push(c); chars.next(); col += 1; }
                        None => return Err(LexError {
                            message: "unterminated string at EOF".into(), line, col }),
                    }
                }
                tokens.push(Token::Str(s));
            }
            '0'..='9' => {
                let mut s = String::new();
                while let Some(&ch) = chars.peek() {
                    if ch.is_ascii_digit() {
                        s.push(ch);
                        chars.next(); col += 1;
                    } else { break; }
                }

                if chars.peek() == Some(&'.') {
                    let mut clone = chars.clone();
                    clone.next();
                    if matches!(clone.peek(), Some(c) if c.is_ascii_digit()) {
                        chars.next(); col += 1;
                        s.push('.');
                        while let Some(&ch) = chars.peek() {
                            if ch.is_ascii_digit() {
                                s.push(ch);
                                chars.next(); col += 1;
                            } else { break; }
                        }
                        let v: f64 = s.parse().map_err(|e| LexError {
                            message: format!("invalid real {s:?}: {e}"),
                            line, col,
                        })?;
                        tokens.push(Token::Real(v));
                        continue;
                    }
                }
                let n: i64 = s.parse().map_err(|e| LexError {
                    message: format!("invalid integer {s:?}: {e}"),
                    line, col,
                })?;
                tokens.push(Token::Int(n));
            }
            'f' if { let mut clone = chars.clone(); clone.next(); clone.peek() == Some(&'"') } => {
                // f"..." format-string literal
                chars.next(); col += 1; // consume 'f'
                chars.next(); col += 1; // consume '"'
                let parts = lex_fstring(&mut chars, &mut line, &mut col)
                    .map_err(|e| LexError { message: e, line, col })?;
                tokens.push(Token::FStr(parts));
            }
            c if is_ident_start(c) => {
                let mut s = String::new();
                while let Some(&ch) = chars.peek() {
                    if is_ident_continue(ch) {
                        s.push(ch);
                        chars.next(); col += 1;
                    } else { break; }
                }
                tokens.push(keyword_or_ident(s));
            }
            '+' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'+') {
                    chars.next(); col += 1;
                    tokens.push(Token::PlusPlus);
                } else {
                    tokens.push(Token::Plus);
                }
            }
            '*' => { chars.next(); col += 1; tokens.push(Token::Star); }
            '/' => { chars.next(); col += 1; tokens.push(Token::Slash); }
            '(' => { chars.next(); col += 1; tokens.push(Token::LParen);   paren_depth += 1; }
            ')' => { chars.next(); col += 1; tokens.push(Token::RParen);   paren_depth = paren_depth.saturating_sub(1); }
            '{' => { chars.next(); col += 1; tokens.push(Token::LBrace);   paren_depth += 1; }
            '}' => { chars.next(); col += 1; tokens.push(Token::RBrace);   paren_depth = paren_depth.saturating_sub(1); }
            '[' => { chars.next(); col += 1; tokens.push(Token::LBracket); paren_depth += 1; }
            ']' => { chars.next(); col += 1; tokens.push(Token::RBracket); paren_depth = paren_depth.saturating_sub(1); }
            '#' => { chars.next(); col += 1; tokens.push(Token::Hash); }
            ',' => { chars.next(); col += 1; tokens.push(Token::Comma); }
            ':' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'=') {
                    chars.next(); col += 1;
                    tokens.push(Token::ColonEq);
                } else {
                    tokens.push(Token::Colon);
                }
            }
            '.' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'.') {
                    chars.next(); col += 1;
                    tokens.push(Token::DotDot);
                } else {
                    tokens.push(Token::Dot);
                }
            }
            '=' => {
                chars.next(); col += 1;

                if chars.peek() == Some(&'>') {
                    chars.next(); col += 1;
                    tokens.push(Token::Implies);
                } else {
                    tokens.push(Token::Eq);
                }
            }
            '<' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'=') {
                    chars.next(); col += 1;
                    tokens.push(Token::Le);
                } else {
                    tokens.push(Token::Lt);
                }
            }
            '>' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'=') {
                    chars.next(); col += 1;
                    tokens.push(Token::Ge);
                } else {
                    tokens.push(Token::Gt);
                }
            }
            '!' => {
                chars.next(); col += 1;
                if chars.peek() == Some(&'=') {
                    chars.next(); col += 1;
                    tokens.push(Token::Neq);
                } else {
                    return Err(LexError { message: "unexpected '!'".into(), line, col });
                }
            }
            '\u{2208}' => { chars.next(); col += 1; tokens.push(Token::In); }
            '\u{2209}' => { chars.next(); col += 1; tokens.push(Token::NotIn); }
            '\u{220B}' => { chars.next(); col += 1; tokens.push(Token::ContainsRev); }
            '\u{2227}' => { chars.next(); col += 1; tokens.push(Token::And); }
            '\u{2228}' => { chars.next(); col += 1; tokens.push(Token::Or); }
            '\u{00AC}' => { chars.next(); col += 1; tokens.push(Token::Not); }
            '\u{21D2}' => { chars.next(); col += 1; tokens.push(Token::Implies); }
            '\u{0394}' => { chars.next(); col += 1; tokens.push(Token::Delta); }
            '\u{2264}' => { chars.next(); col += 1; tokens.push(Token::Le); }
            '\u{2265}' => { chars.next(); col += 1; tokens.push(Token::Ge); }
            '\u{2260}' => { chars.next(); col += 1; tokens.push(Token::Neq); }
            '\u{2200}' => { chars.next(); col += 1; tokens.push(Token::ForAll); }
            '\u{2203}' => { chars.next(); col += 1; tokens.push(Token::Exists); }
            '\u{21A6}' => { chars.next(); col += 1; tokens.push(Token::MapsTo); }
            '\u{00B7}' => { chars.next(); col += 1; tokens.push(Token::MidDot); }
            '\u{00D7}' => { chars.next(); col += 1; tokens.push(Token::Times); }
            '\u{27E8}' => { chars.next(); col += 1; tokens.push(Token::LSeq); paren_depth += 1; }
            '\u{27E9}' => { chars.next(); col += 1; tokens.push(Token::RSeq); paren_depth = paren_depth.saturating_sub(1); }
            '|' => { chars.next(); col += 1; tokens.push(Token::Pipe); }
            '?' => { chars.next(); col += 1; tokens.push(Token::Question); }
            other => {
                return Err(LexError {
                    message: format!("unexpected character {:?}", other),
                    line, col,
                });
            }
        }
        sync_locs!(tok_start);
    }

    tokens.push(Token::Eof);
    sync_locs!((line, col));
    Ok((tokens, locs))
}

/// Lex the body of an f-string after the opening `f"` has been consumed.
/// Handles `{{`/`}}` as escaped literal braces, normal string escapes in
/// literal runs, and `{expr}` interpolation spans. Returns the alternating
/// parts list, or an error message string on malformed input.
fn lex_fstring(
    chars: &mut std::iter::Peekable<std::str::Chars>,
    _line: &mut usize,
    col: &mut usize,
) -> Result<Vec<FStrPart>, String> {
    let mut parts: Vec<FStrPart> = Vec::new();
    let mut lit = String::new();

    loop {
        match chars.peek().copied() {
            None => return Err("unterminated f-string at EOF".into()),
            Some('\n') => return Err("unterminated f-string literal".into()),
            Some('"') => {
                chars.next(); *col += 1;
                // flush any trailing literal
                if !lit.is_empty() {
                    parts.push(FStrPart::Lit(std::mem::take(&mut lit)));
                }
                return Ok(parts);
            }
            Some('{') => {
                chars.next(); *col += 1;
                if chars.peek() == Some(&'{') {
                    // `{{` → escaped literal `{`
                    chars.next(); *col += 1;
                    lit.push('{');
                } else {
                    // start of an interpolation — collect until matching `}`
                    // flush literal so far
                    if !lit.is_empty() {
                        parts.push(FStrPart::Lit(std::mem::take(&mut lit)));
                    }
                    let mut interp = String::new();
                    let mut depth: usize = 1;
                    loop {
                        match chars.peek().copied() {
                            None => return Err("unterminated f-string interpolation at EOF".into()),
                            Some('\n') => return Err("unterminated f-string interpolation".into()),
                            Some('{') => { depth += 1; interp.push('{'); chars.next(); *col += 1; }
                            Some('}') => {
                                chars.next(); *col += 1;
                                depth -= 1;
                                if depth == 0 { break; }
                                interp.push('}');
                            }
                            Some(c) => { interp.push(c); chars.next(); *col += 1; }
                        }
                    }
                    let s = interp.trim().to_string();
                    if s.is_empty() {
                        return Err("empty interpolation `{}` in f-string".into());
                    }
                    parts.push(FStrPart::Interp(s));
                }
            }
            Some('}') => {
                chars.next(); *col += 1;
                if chars.peek() == Some(&'}') {
                    // `}}` → escaped literal `}`
                    chars.next(); *col += 1;
                    lit.push('}');
                } else {
                    return Err("unexpected `}` in f-string (use `}}` for a literal `}`)".into());
                }
            }
            Some('\\') => {
                chars.next(); *col += 1;
                match chars.peek().copied() {
                    Some('"')  => { lit.push('"');  chars.next(); *col += 1; }
                    Some('\\') => { lit.push('\\'); chars.next(); *col += 1; }
                    Some('n')  => { lit.push('\n'); chars.next(); *col += 1; }
                    Some('t')  => { lit.push('\t'); chars.next(); *col += 1; }
                    Some(c)    => return Err(format!("unknown escape \\{} in f-string", c)),
                    None       => return Err("unterminated escape in f-string".into()),
                }
            }
            Some(c) => {
                lit.push(c); chars.next(); *col += 1;
            }
        }
    }
}

fn is_ident_start(c: char) -> bool {
    c.is_ascii_alphabetic() || c == '_'
}

fn is_ident_continue(c: char) -> bool {
    c.is_ascii_alphanumeric() || c == '_'
}

fn keyword_or_ident(s: String) -> Token {
    match s.as_str() {
        "schema"   => Token::Schema,
        "claim"    => Token::Claim,
        "type"     => Token::Type,
        "subclaim" => Token::Subclaim,
        "operator" => Token::Operator,
        "fsm"      => Token::Fsm,
        "fti"      => Token::Fti,
        "external" => Token::External,
        "enum"     => Token::Enum,
        "match"    => Token::Match,
        "matches"  => Token::Matches,
        "import"   => Token::Import,
        "in"       => Token::In,
        "true"     => Token::True,
        "false"    => Token::False,
        "mapsto"   => Token::MapsTo,
        _ => Token::Ident(s),
    }
}

#[cfg(test)]
mod fstr_tests {
    use super::*;

    fn first_fstr(src: &str) -> Vec<FStrPart> {
        let toks = tokenize(src).unwrap();
        toks.into_iter().find_map(|t| match t {
            Token::FStr(parts) => Some(parts),
            _ => None,
        }).expect("no FStr token found")
    }

    #[test]
    fn test_fstr_literal_only() {
        let parts = first_fstr(r#"f"hello""#);
        assert_eq!(parts, vec![FStrPart::Lit("hello".into())]);
    }

    #[test]
    fn test_fstr_interp_only() {
        let parts = first_fstr(r#"f"{a}""#);
        assert_eq!(parts, vec![FStrPart::Interp("a".into())]);
    }

    #[test]
    fn test_fstr_lit_then_interp() {
        let parts = first_fstr(r#"f"v={a}""#);
        assert_eq!(parts, vec![
            FStrPart::Lit("v=".into()),
            FStrPart::Interp("a".into()),
        ]);
    }

    #[test]
    fn test_fstr_escaped_braces() {
        let parts = first_fstr(r#"f"{{lit}}""#);
        assert_eq!(parts, vec![FStrPart::Lit("{lit}".into())]);
    }

    #[test]
    fn test_fstr_empty() {
        let parts = first_fstr(r#"f"""#);
        assert_eq!(parts, vec![]);
    }

    #[test]
    fn test_fstr_multi_interp() {
        let parts = first_fstr(r#"f"{a}-{b}""#);
        assert_eq!(parts, vec![
            FStrPart::Interp("a".into()),
            FStrPart::Lit("-".into()),
            FStrPart::Interp("b".into()),
        ]);
    }
}
